gather.go

```diff
@@ -644,77 +648,207 @@ func (a *Agent) gatherCandidatesRelay(ctx context.Context, urls []*stun.URI) { /
+	relayAllocation, err := client.AllocateTCP()
+	if err != nil {
+		client.Close()
+		closeConnAndLog(locConn, a.log, "failed to allocate on TURN client %s %s", turnServerAddr, err)
+		return
+	}
 
-				client.Close()
-				closeConnAndLog(locConn, a.log, "failed to create relay candidate: %s %s: %v", network, rAddr.String(), err)
-				return
-			}
+	// This `dataConn` is a wrapper for the "data" relay connection we will send/receive data to on
+	// behalf of communicating with a peer.
+	//
+	// Pion uses `net.PacketConn` as the base level for sending data between
+	// peers/relays. PacketConns are unfortunately a UDP abstraction. Therefore pion also has a
+	// `tcpPacketConn` for turning a `net.Conn` into a `net.PacketConn`.
+	//
+	// These `tcpPacketConn`s internally keep a map of TCP connections. We'll only create one entry
+	// in this map (first "caller" wins).
+	dataConn := newTCPPacketConn(tcpPacketParams{
+		ReadBuffer:  1024, // Number of packets?
+		LocalAddr:   locConn.LocalAddr(),
+		Logger:      a.loggerFactory.NewLogger("tcprelay"),
+		WriteBuffer: 4 * 1024 * 1024, // Number of bytes?
+	})
+
+	go func() {
+		// Wait for a `BindIndication`. And follow that by issuing a `ConnectionBind`
+		// request. Returning the `tcpConn` we can to plain old reads/writes.
+		tcpConn, err := relayAllocation.AcceptTCP()
+		if err != nil {
+			panic(err)
+		}
 
-			if err := a.addCandidate(ctx, candidate, relayConn); err != nil {
-				relayConnClose()
+		// Add the minted `tcpConn` to the `dataConn` that the relay candidate will be mapped to.
+		dataConn.AddConn(tcpConn, nil)
+	}()
 
-				if closeErr := candidate.close(); closeErr != nil {
-					a.log.Warnf("Failed to close candidate: %v", closeErr)
-				}
-				a.log.Warnf("Failed to append to localCandidates and run onCandidateHdlr: %v", err)
-			}
-		}(*urls[i])
+	// The `relayAllocation` presumably must be an IP:port value already (as both ends
+	// of a single allocation relay connection must talk to the same relay server). But
+	// just in case we feed the relayAllocation address through a resolution step.
+	relayAddr := relayAllocation.Addr()
+	relayTCPAddr, err := net.ResolveTCPAddr("tcp", relayAddr.String())
+	if err != nil {
+		client.Close()
+		closeConnAndLog(locConn, a.log, "failed to resolve relay address %s %s", relayAddr, err)
+		return
+	}
+
+	relayConfig := CandidateRelayConfig{
+		Network:         network,
+		Component:       ComponentRTP,
+		Address:         relayTCPAddr.IP.String(),
+		Port:            relayTCPAddr.Port,
+		RelAddr:         relatedAddr,
+		RelPort:         relatedPort,
+		TURNControlConn: client,
+		// Dan: We force TCPTypePassive here by choice. Maybe SimultaneousOpen is valid as well? We
+		// certainly expect to only succeed via the other peer dialing out to the relay. But if we
+		// have firewall/nat limitations on the TURN server, simultaneous open may have an
+		// advantage? Can certainly experiment here at a later date.
```

Comment:
Yeah I don't understand how SO works. If we're not dealing with a relay server in between, and both sides open a connection to the other side, how do they agree on which connection to actually use?

Reply:
Yeah I have similar concerns. pion/ice has a whole "tcp packet conn" abstraction which has a map of sockets under the hood. I suspect the (expected) answer lies there.

However, I have no confidence the solution is coded up and works.